### PR TITLE
Update "Getting started" and "api" sections of the documentation

### PR DESCRIPTION
**Pull request by phackstock**
After the refactoring in #73, I've now updated the getting started and api sections.

Next up will be the usage section but I'd prefer to keep those PRs separate.

### CODE REVIEW

File: doc/source/api.rst [hunk 1]
```diff
@@ -5,6 +5,14 @@
 API documentation
 =================
 
+.. _process-function:
+
+Process function
```
Comment (danielhuppmann) [inline doc/source/api.rst:10]:
```suggestion
Top-level functions
```

File: doc/source/getting_started.rst [hunk 2]
```diff
@@ -90,35 +92,34 @@ The following outlines how to use the nomenclature package:
...
-   dsd.validate(df)
-   # Apply region processing to the data
-   # Only to be used if there are mappings!
-   df = rp.apply(df)
+  # Import the necessary libraries
+  import pyam
+  from nomenclature import DataStructureDefinition, RegionProcessor, process
+  
+  # Initialize DataStructureDefinition and RegionProcessor giving them the
+  # directories where the codelists and mappings are defined as input.
```
Comment (danielhuppmann) [inline doc/source/getting_started.rst:100]:
```suggestion
  # Initialize DataStructureDefinition from a suitable directory
```

File: doc/source/getting_started.rst [hunk 3]
```diff
@@ -90,35 +92,34 @@ The following outlines how to use the nomenclature package:
...
+  dsd = DataStructureDefinition("definitions/")
```
Comment (danielhuppmann) [inline doc/source/getting_started.rst:101]:
```suggestion
  dsd = DataStructureDefinition("definitions")
```

File: doc/source/getting_started.rst [hunk 4]
```diff
@@ -90,35 +92,34 @@ The following outlines how to use the nomenclature package:
...
+  
+  # Only to be used if there are mappings!
+  rp = RegionProcessor.from_directory("mappings/", dsd)
```
Comment (danielhuppmann) [inline doc/source/getting_started.rst:104]: The `dsd` argument was removed, right?
```suggestion
  rp = RegionProcessor.from_directory("mappings")
```

File: doc/source/getting_started.rst [hunk 5]
```diff
@@ -90,35 +92,34 @@ The following outlines how to use the nomenclature package:
...
+  
+  # Read the data using pyam
+  iam_results_file = "some file"
+  df = pyam.IamDataFrame(iam_results_file)
```
Comment (danielhuppmann) [inline doc/source/getting_started.rst:108]:
```suggestion
  df = pyam.IamDataFrame("/path/to/file")
```

File: doc/source/getting_started.rst [hunk 6]
```diff
@@ -90,35 +92,34 @@ The following outlines how to use the nomenclature package:
...
+  
+  # Using the process function we perform the validation of allowed regions and
+  # variables as well as the aggregation (if applicable) in one step.
```
Comment (danielhuppmann) [inline doc/source/getting_started.rst:111]:
```suggestion
  # Perform the validation and apply the region aggregation
```

File: doc/source/api.rst [hunk 7]
```diff
@@ -5,6 +5,14 @@
 API documentation
 =================
 
+.. _process-function:
```
Comment (danielhuppmann) [inline doc/source/api.rst:8]:
```suggestion
.. _toplevel-functions:
```

File: doc/source/getting_started.rst [hunk 8]
```diff
@@ -90,35 +92,34 @@ The following outlines how to use the nomenclature package:
...
   containing yaml files. See :ref:`dir-structure` for details. 
 
-* *DataStructureDefinition.apply()* returns *None* if the data frame only contains   
-  allowed values and raises an error otherwise.
+* The processor argument of *process* is optional and may only to be used if there are  
+  model mappings. See :ref:`process-function` for details.
 
-* *RegionProcessor* is only to be used if there are model mappings.
+* If not all dimensions of the **DataStructureDefinition** should be validated, a
+  *dimensions* argument in form of a list of strings can be provided. Only the provided dimensions will then be validated. See :ref:`process-function` for details.
```
Comment (danielhuppmann) [inline doc/source/getting_started.rst:125]:
```suggestion
  *dimensions* argument in form of a list of strings can be provided. Only the provided dimensions will then be validated. See :func:`nomenclature.process` for details.
```

File: doc/source/getting_started.rst [hunk 9]
```diff
@@ -90,35 +92,34 @@ The following outlines how to use the nomenclature package:
...
-   df = rp.apply(df)
+  # Import the necessary libraries
+  import pyam
+  from nomenclature import DataStructureDefinition, RegionProcessor, process
+  
+  # Initialize DataStructureDefinition and RegionProcessor giving them the
+  # directories where the codelists and mappings are defined as input.
+  dsd = DataStructureDefinition("definitions/")
+  
+  # Only to be used if there are mappings!
```
Comment (danielhuppmann) [inline doc/source/getting_started.rst:103]:
```suggestion
  # Initialize a RegionProcessor from a suitable directory that has the mappings
```

File: nomenclature/core.py [hunk 10]
```diff
@@ -8,11 +10,49 @@
...
-    processor: RegionProcessor = None,
-):
+    dimensions: Optional[List[str]] = None,
+    processor: Optional[RegionProcessor] = None,
+) -> pyam.IamDataFrame:
+    """Function for validation and region aggregation in one step
+
+    Since there are some subtleties in the order of applying region processing and
+    validation, this function is the recommended way of using the nomenclature package
+    as it takes care of subtleties.
```
Comment (danielhuppmann) [inline nomenclature/core.py:22]:
```suggestion
   This function is the recommended way of using the nomenclature package
   for validation and region-processing.
```

File: nomenclature/core.py [hunk 11]
```diff
@@ -8,11 +10,49 @@
...
+    df : pyam.IamDataFrame
+        Input data to be validated and aggregated.
+    dsd : DataStructureDefinition
+        Data templates that are used for validation.
+    dimensions : Optional[List[str]], optional
+        List of dimensions that will be used in the validation, if None all
+        dsd.dimensions will be used, by default None
+    processor : Optional[RegionProcessor], optional
+        Region processor that will perform region renaming and aggregation if provided,
+        by default None
```
Comment (danielhuppmann) [inline nomenclature/core.py:35]: The "by default None" can be seen from the function signature no need to repeat
```suggestion
        Region processor that will perform region renaming and aggregation if provided
```

File: nomenclature/core.py [hunk 12]
```diff
@@ -8,11 +10,49 @@
...
+
+    Parameters
+    ----------
+    df : pyam.IamDataFrame
+        Input data to be validated and aggregated.
+    dsd : DataStructureDefinition
+        Data templates that are used for validation.
+    dimensions : Optional[List[str]], optional
+        List of dimensions that will be used in the validation, if None all
+        dsd.dimensions will be used, by default None
```
Comment (danielhuppmann) [inline nomenclature/core.py:32]: No need to repeat "by default None"
```suggestion
        Dimensions to be used in the validation, defaults to all dimensions defined in *dsd* 
```

File: nomenclature/core.py [hunk 13]
```diff
@@ -8,11 +10,49 @@
...
+    validation, this function is the recommended way of using the nomenclature package
+    as it takes care of subtleties.
+
+    Parameters
+    ----------
+    df : pyam.IamDataFrame
+        Input data to be validated and aggregated.
+    dsd : DataStructureDefinition
+        Data templates that are used for validation.
+    dimensions : Optional[List[str]], optional
```
Comment (danielhuppmann) [inline nomenclature/core.py:30]:
```suggestion
    dimensions : list, optional
```

File: nomenclature/core.py [hunk 14]
```diff
@@ -8,11 +10,49 @@
...
+    Since there are some subtleties in the order of applying region processing and
+    validation, this function is the recommended way of using the nomenclature package
+    as it takes care of subtleties.
+
+    Parameters
+    ----------
+    df : pyam.IamDataFrame
+        Input data to be validated and aggregated.
+    dsd : DataStructureDefinition
+        Data templates that are used for validation.
```
Comment (danielhuppmann) [inline nomenclature/core.py:29]: The term "data template" is used on the index page but nowhere else - we should aim to use consistent terms to describe the classes and use cases

File: nomenclature/core.py [hunk 15]
```diff
@@ -8,11 +10,49 @@
...
+        Processed data frame
+
+    Notes
+    -----
+    The above mentioned "subtleties" in the order of operations is related to the fact
+    that as part of the region processing, three things can occur:
+
+    1. Model native regions not mentioned in the model mapping will be dropped
+    2. Model native regions can be renamed
+    3. Common regions based on aggregated model native regions weill be newly created.
```
Comment (danielhuppmann) [inline nomenclature/core.py:49]:
```suggestion
    3. Aggregation from model native regions to "common regions"
```

**Review by danielhuppmann — COMMENTED** [review]
The documentation would benefit, in my opinion, from a more direct, less verbose writing style. See some suggestions inline.

File: doc/source/getting_started.rst (hunk 8 above)
Comment (phackstock) [reply doc/source/getting_started.rst:125]: Ah that's how you link functions directly, great, thanks.

File: nomenclature/core.py [hunk 16]
```diff
@@ -8,11 +10,49 @@
...
+
+    Returns
+    -------
+    pyam.IamDataFrame
+        Processed data frame
+
+    Notes
+    -----
+    The above mentioned "subtleties" in the order of operations is related to the fact
+    that as part of the region processing, three things can occur:
```
[the remark this answers is not in the document]
Comment (phackstock) [reply nomenclature/core.py:45]: Should I leave this section out completely then?

[the remark this answers is not in the document]
Comment (danielhuppmann) [reply nomenclature/core.py:42]: Repeating what the function does makes sense - maybe shorten it as suggested and put that in the description at the top of the docstring?

[the remark this answers is not in the document]
Comment (phackstock) [reply nomenclature/core.py:42]: Ok will do.

File: nomenclature/core.py (hunk 14 above)
Comment (phackstock) [reply nomenclature/core.py:28]: What's a better term then? Should I just change it to 'Used for validation' (leave out the "data templates")?

Comment (danielhuppmann) [reply nomenclature/core.py:28]: The docstring of the class uses "Definition of datastructure codelists". Maybe use
> Definitions of codelists (usually project-specific) used for validation

Comment (phackstock) [reply nomenclature/core.py:28]: Perfect, thanks, I'll use this. I'll also update it in the index page for consistency.

File: doc/source/getting_started.rst [hunk 17]
```diff
@@ -29,9 +29,11 @@ attributes (e.g., description, unit, parent region).
 A **RegionAggregationMapping** is a mapping that defines on a per-model basis how model
 native regions should be renamed and aggregated to comparison regions.
 
-
-Before the :ref:`minimum-working-example` can be covered, the required directory
-structure for definitions and mappings needs to be addressed. 
+In order to reduce the amount of code required, there is a convenience function called
```
[the remark this answers is not in the document]
Comment (phackstock) [reply doc/source/getting_started.rst]: Done

File: nomenclature/core.py (hunk 15 above)
Comment (phackstock) [reply nomenclature/core.py]: Done

File: nomenclature/core.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (phackstock) [reply nomenclature/core.py]: Done.

**Comment by phackstock** [comment]
Shortened the docstring of process and the getting started section.

**Review by danielhuppmann — APPROVED** [review]
Thanks @phackstock, nice improvement!

**Comment by phackstock** [comment]
Perfect, now just the usage part missing.

**Pull request merged**